File: 70.climbing-stairs.js
/*
 * @lc app=leetcode id=70 lang=javascript
 *
 * [70] Climbing Stairs
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number}
 */
var climbStairs = function (n, m = new Map()) {
  if (n <= 2) return n;
  if (m.has(n)) return m.get(n);

  const solutinon = climbStairs(n - 1, m) + climbStairs(n - 2, m);
  m.set(n, solutinon);
  return solutinon;
};
// @lc code=end

File: solutions/70.climbing-stairs.rb
#
# @lc app=leetcode id=70 lang=ruby
#
# [70] Climbing Stairs
#

# @lc code=start
# @param {Integer} n
# @return {Integer}
def climb_stairs(n)
  return n if n <= 2

  dp = [1, 1, 2]
  (3..n).each do |i|
    dp[i] = dp[i - 1] + dp[i - 2]
  end
  dp[n]
end
# @lc code=end
